rclpy/rclpy/qos_event.py

```diff
@@ -0,0 +1,172 @@
+
+Mirrors rmw_liveliness_changed_status_t from rmw/types.h
+"""
+QoSLivelinessChangedInfo = namedtuple(
+    'QoSLivelinessChangedInfo',
+    ('alive_count', 'not_alive_count', 'alive_count_change', 'not_alive_count_change'))
+
+"""
+Payload type for Publisher Deadline callback.
+
+Mirrors rmw_offered_deadline_missed_status_t from rmw/types.h
+"""
+QoSOfferedDeadlineMissedInfo = namedtuple(
+    'QoSOfferedDeadlineMissedInfo',
+    ('total_count', 'total_count_change'))
+
+"""
+Payload type for Publisher Liveliness callback.
+
+Mirrors rmw_liveliness_lost_status_t from rmw/types.h
+"""
+QoSLivelinessLostInfo = namedtuple(
+    'QoSLivelinessLostInfo',
+    ('total_count', 'total_count_change'))
+
+
+class SubscriptionEventCallbacks:
+    """Container to provide middleware event callbacks for a Subscription."""
+
+    def __init__(
+        self,
+        *,
+        deadline: Callable[[QoSRequestedDeadlineMissedInfo], None] = None,
+        liveliness: Callable[[QoSLivelinessChangedInfo], None] = None,
+    ) -> None:
+        """
+        Constructor.
+
+        :param deadline: A user-defined callback that is called when a topic misses our
+            requested Deadline.
+        :param liveliness: A user-defined callback that is called when the Liveliness of
+            a Publisher on subscribed topic changes.
+        """
+        self.deadline = deadline
+        self.liveliness = liveliness
+
+
+class PublisherEventCallbacks:
+    """Container to provide middleware event callbacks for a Publisher."""
+
+    def __init__(
+        self,
+        *,
+        deadline: Callable[[QoSOfferedDeadlineMissedInfo], None] = None,
+        liveliness: Callable[[QoSLivelinessLostInfo], None] = None
+    ) -> None:
+        """
+        Constructor.
+
+        :param deadline: A user-defined callback that is called when the Publisher misses
+            its offered Deadline.
+        :param liveliness: A user-defined callback that is called when this Publisher
+            fails to signal its Liveliness and is reported as not-alive.
+        """
+        self.deadline = deadline
+        self.liveliness = liveliness
+
+
+class QoSEventHandler(Waitable):
+    """Waitable type to handle QoS events."""
+
+    def __init__(
+        self,
+        *,
+        callback_group,
+        callback,
+        event_type,
+        parent_handle,
+    ):
+        # Waitable init adds self to callback_group
+        super().__init__(callback_group)
+        self.event_type = event_type
+        self.callback = callback
+
+        self._parent_handle = parent_handle
+        self._event_handle = _rclpy.rclpy_create_event(event_type, parent_handle)
+        self._event_is_ready = False
+        self._event_index = None
+
+    # Start Waitable API
+    def is_ready(self, wait_set):
+        """Return True if entities are ready in the wait set."""
+        if _rclpy.rclpy_wait_set_is_ready('event', wait_set, self._event_index):
+            self._event_is_ready = True
```

Comment:
I don't fully understand the code but it seems weird this `is_ready` method is flipping this boolean value.
Events aren't ready until the user checks for them?
This may just be a doc issue.

Reply:
The Python-side event handler doesn't have ready events until we check for them - yeah. I was copying existing conventions like

https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/action/client.py#L223
and
https://github.com/ros2/rclpy/blob/master/rclpy/test/test_waitable.py#L55